File: flextarget/Model/ScoringUtility.swift
import Foundation
import CoreData
import SwiftUI

/// Utility class for drill scoring calculations
class ScoringUtility {
    
    /// Calculate score for a specific hit area
    static func scoreForHitArea(_ hitArea: String) -> Int {
        let trimmed = hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        
        switch trimmed {
        case "azone":
            return 5
        case "czone":
            return 3
        case "dzone":
            return 2
        case "miss":
            return -15
        case "whitezone":
            return -25
        case "blackzone":
            return -10
        case "circlearea": // Paddle
            return 5
        case "popperzone": // Popper
            return 5
        default:
            return 0
        }
    }
    
    /// Calculate the number of missed targets
    static func calculateMissedTargets(shots: [ShotData], drillSetup: DrillSetup?) -> Int {
        guard let targetsSet = drillSetup?.targets as? Set<DrillTargetsConfig> else {
            return 0
        }
        
        let expectedTargets = Set(targetsSet.map { $0.targetName ?? "" }.filter { !$0.isEmpty })
        let shotsDevices = Set(shots.compactMap { $0.device ?? $0.target })
        
        let missedTargets = expectedTargets.subtracting(shotsDevices)
        return missedTargets.count
    }
    
    /// Calculate total score with drill rules applied
    static func calculateTotalScore(shots: [ShotData], drillSetup: DrillSetup?) -> Double {
        // Group shots by target/device
        var shotsByTarget: [String: [ShotData]] = [:]
        for shot in shots {
            let device = shot.device ?? shot.target ?? "unknown"
            if shotsByTarget[device] == nil {
                shotsByTarget[device] = []
            }
            shotsByTarget[device]?.append(shot)
        }
        
        // Keep best 2 shots per target, but always include no-shoot zone hits
        // Exception: for paddle and popper targets, keep all shots (no best 2 limit)
        var bestShotsPerTarget: [ShotData] = []
        for (_, targetShots) in shotsByTarget {
            // Detect target type from shots
            let targetType = targetShots.first?.content.targetType.lowercased() ?? ""
            let isPaddleOrPopper = targetType == "paddle" || targetType == "popper"
            
            let noShootZoneShots = targetShots.filter { shot in
                let trimmed = shot.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                return trimmed == "whitezone" || trimmed == "blackzone"
            }
            
            let otherShots = targetShots.filter { shot in
                let trimmed = shot.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                return trimmed != "whitezone" && trimmed != "blackzone"
            }
            
            // For paddle and popper: keep all shots; for others: keep best 2
            let selectedOtherShots: [ShotData]
            if isPaddleOrPopper {
                let validShots = otherShots.filter { s in
                    let a = s.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                    return a != "miss" && a != "m" && !a.isEmpty
                }
                if validShots.count >= 2 {
                    selectedOtherShots = validShots
                } else {
                    selectedOtherShots = otherShots
                }
            } else {
                let sortedOtherShots = otherShots.sorted {
                    Double(ScoringUtility.scoreForHitArea($0.content.hitArea)) > Double(ScoringUtility.scoreForHitArea($1.content.hitArea))
                }
                selectedOtherShots = Array(sortedOtherShots.prefix(2))
            }
            
            // Always include no-shoot zone shots
            bestShotsPerTarget.append(contentsOf: noShootZoneShots)
            bestShotsPerTarget.append(contentsOf: selectedOtherShots)
        }
        
        var totalScore = bestShotsPerTarget.reduce(0.0) { $0 + Double(ScoringUtility.scoreForHitArea($1.content.hitArea)) }
        
        // Auto re-evaluate score: deduct 10 points for each missed target
        let missedTargetCount = ScoringUtility.calculateMissedTargets(shots: shots, drillSetup: drillSetup)
        let missedTargetPenalty = missedTargetCount * 10
        totalScore -= Double(missedTargetPenalty)
        
        // Ensure score never goes below 0
        return max(0, totalScore)
    }
    
    /// Calculate effective hit zone counts based on drill rules
    static func calculateEffectiveCounts(shots: [ShotData]) -> [String: Int] {
        // Group shots by target/device
        var shotsByTarget: [String: [ShotData]] = [:]
        for shot in shots {
            let device = shot.device ?? shot.target ?? "unknown"
            if shotsByTarget[device] == nil {
                shotsByTarget[device] = []
            }
            shotsByTarget[device]?.append(shot)
        }
        
        var aCount = 0
        var cCount = 0
        var dCount = 0
        var nCount = 0
        var mCount = 0
        
        for (_, targetShots) in shotsByTarget {
            // Detect target type from shots
            let targetType = targetShots.first?.content.targetType.lowercased() ?? ""
            let isPaddleOrPopper = targetType == "paddle" || targetType == "popper"
            
            let noShootZoneShots = targetShots.filter { shot in
                let trimmed = shot.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                return trimmed == "whitezone" || trimmed == "blackzone"
            }
            
            let otherShots = targetShots.filter { shot in
                let trimmed = shot.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                return trimmed != "whitezone" && trimmed != "blackzone"
            }
            
            // Count no-shoot zones (always included)
            nCount += noShootZoneShots.count
            
            // For paddle and popper: count all scoring shots; for others: count best 2
            let scoringShots: [ShotData]
            
            let validShots = otherShots.filter { s in
                let a = s.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                return a != "miss" && a != "m" && !a.isEmpty
            }
            
            if isPaddleOrPopper {
                if validShots.count >= 2 {
                    scoringShots = validShots
                } else {
                    scoringShots = otherShots
                }
            } else {
                let sortedOtherShots = otherShots.sorted {
                    Double(ScoringUtility.scoreForHitArea($0.content.hitArea)) > Double(ScoringUtility.scoreForHitArea($1.content.hitArea))
                }
                scoringShots = Array(sortedOtherShots.prefix(2))
            }
            
            // Count effective shots by zone
            for shot in scoringShots {
                let area = shot.content.hitArea.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                switch area {
                case "azone":
                    aCount += 1
                case "czone":
                    cCount += 1
                case "dzone":
                    dCount += 1
                case "miss", "m", "":
                    // Special consideration: if target has 2 valid shots, don't count miss
                    // (This is already handled for non-paddles by prefix(2), 
                    // and for paddles by the validShots.count >= 2 check above)
                    if validShots.count < 2 {
                        mCount += 1
                    }
                default:
                    break
                }
            }
        }
        
        return ["A": aCount, "C": cCount, "D": dCount, "N": nCount, "M": mCount]
    }
    
    /// Calculate score based on adjusted hit zone metrics
    static func calculateScoreFromAdjustedHitZones(_ adjustedHitZones: [String: Int]?, drillSetup: DrillSetup?) -> Int {
        guard let adjustedHitZones = adjustedHitZones else { return 0 }
        
        let aCount = adjustedHitZones["A"] ?? 0
        let cCount = adjustedHitZones["C"] ?? 0
        let dCount = adjustedHitZones["D"] ?? 0
        let nCount = adjustedHitZones["N"] ?? 0  // No-shoot zones
        let peCount = adjustedHitZones["PE"] ?? 0  // Penalty count
        let mCount = adjustedHitZones["M"] ?? 0
        
        // Calculate base score from adjusted counts
        // A=5, C=3, D=2, N=-10 (black zone) or -25 (white zone), M=-15
        var totalScore = (aCount * 5) + (cCount * 3) + (dCount * 2) + (mCount * -15)
        
        // Apply penalties for no-shoot zones
        // Assume half are black zone (-10) and half are white zone (-25) for average penalty
        let avgNoShootPenalty = nCount > 0 ? (nCount / 2 * (-10) + (nCount + 1) / 2 * (-25)) : 0
        totalScore += avgNoShootPenalty
        
        // Apply penalty deductions (10 points per PE)
        let penaltyDeduction = peCount * 10
        totalScore -= penaltyDeduction
        
        // Ensure score never goes below 0
        return max(0, totalScore)
    }
    
}

File: FlexTargetAndroid/app/src/main/java/com/flextarget/android/data/model/ScoringUtility.kt
package com.flextarget.android.data.model

import com.flextarget.android.data.local.entity.DrillSetupEntity
import com.flextarget.android.data.local.entity.DrillTargetsConfigEntity

/**
 * Utility class for drill scoring calculations
 * Ported from iOS ScoringUtility
 */
object ScoringUtility {

    /**
     * Calculate score for a specific hit area
     */
    fun scoreForHitArea(hitArea: String): Int {
        val trimmed = hitArea.trim().lowercase()
        return when (trimmed) {
            "azone" -> 5
            "czone" -> 3
            "dzone" -> 2
            "miss" -> -15
            "whitezone" -> -25
            "blackzone" -> -10
            "circlearea" -> 5 // Paddle
            "popperzone" -> 5 // Popper
            else -> 0
        }
    }

    /**
     * Calculate the number of missed targets
     */
    fun calculateMissedTargets(shots: List<ShotData>, targets: List<DrillTargetsConfigEntity>?): Int {
        val targetsSet = targets ?: return 0
        val expectedTargets = targetsSet.mapNotNull { it.targetName }.filter { it.isNotEmpty() }.toSet()
        val shotsDevices = shots.mapNotNull { it.device ?: it.target }.toSet()
        val missedTargets = expectedTargets.subtract(shotsDevices)
        return missedTargets.size
    }

    /**
     * Calculate total score with drill rules applied
     */
    fun calculateTotalScore(shots: List<ShotData>, targets: List<DrillTargetsConfigEntity>?): Double {
        // Group shots by target/device
        val shotsByTarget = mutableMapOf<String, MutableList<ShotData>>()
        for (shot in shots) {
            val device = shot.device ?: shot.target ?: "unknown"
            shotsByTarget.getOrPut(device) { mutableListOf() }.add(shot)
        }

        // Keep best 2 shots per target, but always include no-shoot zone hits
        // Exception: for paddle and popper targets, keep all shots (no best 2 limit)
        val bestShotsPerTarget = mutableListOf<ShotData>()
        for ((_, targetShots) in shotsByTarget) {
            // Detect target type from shots
            val targetType = targetShots.firstOrNull()?.content?.actualTargetType?.lowercase() ?: ""
            val isPaddleOrPopper = targetType == "paddle" || targetType == "popper"

            val noShootZoneShots = targetShots.filter { shot ->
                val trimmed = shot.content.actualHitArea.trim().lowercase()
                trimmed == "whitezone" || trimmed == "blackzone"
            }

            val otherShots = targetShots.filter { shot ->
                val trimmed = shot.content.actualHitArea.trim().lowercase()
                trimmed != "whitezone" && trimmed != "blackzone"
            }

            // For paddle and popper: keep all shots; for others: keep best 2
            val selectedOtherShots = if (isPaddleOrPopper) {
                otherShots
            } else {
                val sortedOtherShots = otherShots.sortedByDescending { scoreForHitArea(it.content.actualHitArea) }
                sortedOtherShots.take(2)
            }

            // Always include no-shoot zone shots
            bestShotsPerTarget.addAll(noShootZoneShots)
            bestShotsPerTarget.addAll(selectedOtherShots)
        }

        var totalScore = bestShotsPerTarget.sumOf { scoreForHitArea(it.content.actualHitArea).toDouble() }

        // Auto re-evaluate score: deduct 10 points for each missed target
        val missedTargetCount = calculateMissedTargets(shots, targets)
        val missedTargetPenalty = missedTargetCount * 10
        totalScore -= missedTargetPenalty.toDouble()

        // Ensure score never goes below 0
        return maxOf(0.0, totalScore)
    }
}